lib/compat/wordpress-6.5/class-wp-script-modules.php

```diff
@@ -0,0 +1,359 @@
+	 *                                                                                 query string for cache busting
+	 *                                                                                 purposes. If $version is set to
+	 *                                                                                 false, the version number is the
+	 *                                                                                 currently installed WordPress
+	 *                                                                                 version. If $version is set to
+	 *                                                                                 null, no version is added.
+	 */
+	public function enqueue( $module_id, $src = '', $deps = array(), $version = false ) {
+		if ( isset( $this->registered[ $module_id ] ) ) {
+			$this->registered[ $module_id ]['enqueue'] = true;
+		} elseif ( $src ) {
+			$this->register( $module_id, $src, $deps, $version );
+			$this->registered[ $module_id ]['enqueue'] = true;
+		} else {
+			$this->enqueued_before_registered[ $module_id ] = true;
+		}
+	}
+
+	/**
+	 * Unmarks the module so it will no longer be enqueued in the page.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $module_id The identifier of the module.
+	 */
+	public function dequeue( $module_id ) {
+		if ( isset( $this->registered[ $module_id ] ) ) {
+			$this->registered[ $module_id ]['enqueue'] = false;
+		}
+		unset( $this->enqueued_before_registered[ $module_id ] );
+	}
+
+	/**
+	 * Adds the hooks to print the import map, enqueued modules and module
+	 * preloads.
+	 *
+	 * It adds the actions to print the enqueued modules and module preloads to
+	 * both `wp_head` and `wp_footer` because in classic themes, the modules
+	 * used by the theme and plugins will likely be able to be printed in the
+	 * `head`, but the ones used by the blocks will need to be enqueued in the
+	 * `footer`.
+	 *
+	 * As all modules are deferred and dependencies are handled by the browser,
+	 * the order of the modules is not important, but it's still better to print
+	 * the ones that are available when the `wp_head` is rendered, so the browser
+	 * starts downloading those as soon as possible.
+	 *
+	 * The import map is also printed in the footer to be able to include the
+	 * dependencies of all the modules, including the ones printed in the footer.
+	 *
+	 * @since 6.5.0
+	 */
+	public function add_hooks() {
+		add_action( 'wp_head', array( $this, 'print_import_map' ) );
+		add_action( 'wp_head', array( $this, 'print_enqueued_modules' ) );
+		add_action( 'wp_head', array( $this, 'print_module_preloads' ) );
+		if ( ! wp_is_block_theme() ) {
+			add_action( 'wp_footer', array( $this, 'print_import_map' ) );
+		}
+		add_action( 'wp_footer', array( $this, 'print_module_preloads' ) );
+		add_action( 'wp_footer', array( $this, 'print_enqueued_modules' ) );
+	}
```

Comment:
This differs from Core commit. Interactivity was not working in Classic Themes.
TODO: Test in Core only.
cc: @luisherranz

Reply:
You can't add two import maps.

Reply:
Should we make it conditional then? Otherwise won't work on classic themes.

Reply:
Why do you think the WP Core version won't work on classic themes? It adds the block modules and the import map to the footer:

```php
public function add_hooks() {
  add_action( 'wp_head', array( $this, 'print_enqueued_modules' ) );
  add_action( 'wp_head', array( $this, 'print_module_preloads' ) );
  add_action( 'wp_footer', array( $this, 'print_enqueued_modules' ) );
  add_action( 'wp_footer', array( $this, 'print_module_preloads' ) );
  add_action( 'wp_footer', array( $this, 'print_import_map' ) );
}
```

Reply:
This is the explanation:

```php
/**
 * Adds the hooks to print the import map, enqueued modules and module
 * preloads.
 *
 * It adds the actions to print the enqueued modules and module preloads to
 * both `wp_head` and `wp_footer` because in classic themes, the modules
 * used by the theme and plugins will likely be able to be printed in the
 * `head`, but the ones used by the blocks will need to be enqueued in the
 * `footer`.
 *
 * As all modules are deferred and dependencies are handled by the browser,
 * the order of the modules is not important, but it's still better to print
 * the ones that are available when the `wp_head` is rendered, so the browser
 * starts downloading those as soon as possible.
 *
 * The import map is also printed in the footer to be able to include the
 * dependencies of all the modules, including the ones printed in the footer.
 *
 * @since 6.5.0
 */
```

Reply:
More info about the reasoning in this comment:

- https://github.com/WordPress/wordpress-develop/pull/5818#discussion_r1443043837

Reply:
I haven't tested it out, though. So it's all theoretical.

Reply:
I did not try in Core only yet, but at least, in Gutenberg, having the import map after the enqueued modules fires this error:
![Screenshot 2024-01-12 at 11 06 13](https://github.com/WordPress/gutenberg/assets/37012961/9e3e6653-8aa4-4c91-b499-47bd91aa7a11)

It seems that the import map must be before the rest of scripts. Something like:
```
public function add_hooks() {
		if ( ! wp_is_block_theme() ) {
			add_action( 'wp_footer', array( $this, 'print_import_map' ) );
		} else {
			add_action( 'wp_head', array( $this, 'print_import_map' ) );
		}
		add_action( 'wp_head', array( $this, 'print_enqueued_modules' ) );
		add_action( 'wp_head', array( $this, 'print_module_preloads' ) );
		add_action( 'wp_footer', array( $this, 'print_module_preloads' ) );
		add_action( 'wp_footer', array( $this, 'print_enqueued_modules' ) );
	}
```

Reply:
You're right, this doesn't work:

```html
<!DOCTYPE html>
<html>
  <head>
    <script type="module">
      import lodash from "lodash";
      console.log(lodash.add(1, 2));
    </script>
    <script type="importmap">
      {
        "imports": {
          "lodash": "https://ga.jspm.io/npm:lodash@4.17.21/lodash.js"
        }
      }
    </script>
  </head>
  <body></body>
</html>
```

And this works:

```html
<!DOCTYPE html>
<html>
  <head>
   <script type="importmap">
      {
        "imports": {
          "lodash": "https://ga.jspm.io/npm:lodash@4.17.21/lodash.js"
        }
      }
    </script> <script type="module">
      import lodash from "lodash";
      console.log(lodash.add(1, 2));
    </script>
  </head>
  <body></body>
</html>
```

That's quite a surprise, taking into account that modules are always deferred.

I'll investigate to see if there's any workaround.

Reply:
There doesn't seem to be, so we need to modify that part. Let's do it here first, and then once we've confirmed that it works, submit the patch to Core.

Reply:
I'll test if we can at least still preload them in the head to start downloading them as soon as possible.

Reply:
Oh my, that also fails 🤦‍♂️

Reply:
We need to move everything to the footer.

Reply:
> That's quite a surprise, taking into account that modules are always deferred.

That is surprising. I consulted Bard about this, and it says:

> * While modules are executed in a deferred manner, the process of resolving imports happens earlier.
> * When the browser encounters a module script, it immediately starts parsing it to identify imports.
> * If it doesn't have an import map available yet, it won't be able to determine where to fetch those imports from.

That makes sense to save time so module dependencies can be loaded as soon as possible. I wasn't aware that imports start to get resolved before the execution happens. I double checked the spec and it seems to [confirm this](https://html.spec.whatwg.org/multipage/scripting.html#script-processing-src-prepare), when a module script is encountered:

> 8\. Set el's [delaying the load event](https://html.spec.whatwg.org/multipage/scripting.html#concept-script-delay-load) to true.
> [...]
> 11\. [...] [Fetch an external module script graph](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-module-script-tree) given url, settings object, options, and onComplete.
> [...]
> For performance reasons, user agents may start fetching the classic script or module graph (as defined above) as soon as the [src](https://html.spec.whatwg.org/multipage/scripting.html#attr-script-src) attribute is set, instead, in the hope that el will be inserted into the document

Fetching the external module script graph as early as possible means that the script can execute without delay once the DOM has loaded and the deferred module script is executed. In order to properly resolve the fetching of the external module script graph, the `importmap` would have to be already consumed.

Reply:
> That is surprising.

Haha, the fact that this was also a surprise for you is a bit of a relief, thank you 😄

I guess they did it to _force_ people to do it that way because it is slightly more performant. But it kind of goes against sites that discover their dependencies during the rendering while they are already streaming the HTML because it forces them to push everything till the end. Like the heavily componentized React sites.

Reply:
This could be a performance optimization that could be applied via Core-43258: if pages are output-buffered by default, then the import map could be moved to the `head` even if it is printed in the footer.

Reply:
Ok. But for now, we don't need to do anything special here, right? We just go back to this:

```js
public function add_hooks() {
  $position = wp_is_block_theme() ? 'wp_head' : 'wp_footer';
  add_action( $position, array( $this, 'print_enqueued_modules' ) );
  add_action( $position, array( $this, 'print_module_preloads' ) );
  add_action( $position, array( $this, 'print_import_map' ) );
}
```

If you confirm it, I'll prepare the PR.

Reply:
I'm afraid `print_import_map` needs to go before. I updated this PR. The code should be:
```php
public function add_hooks() {
    $position = wp_is_block_theme() ? 'wp_head' : 'wp_footer';
    add_action( $position, array( $this, 'print_import_map' ) );
    add_action( $position, array( $this, 'print_enqueued_modules' ) );
    add_action( $position, array( $this, 'print_module_preloads' ) );
}
```

Reply:
I opened a PR in WordPress Core and added the code to this PR.

- https://github.com/WordPress/wordpress-develop/pull/5931